src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs

```diff
@@ -602,15 +613,51 @@ public class OutputCacheFilter : FilterProvider, IActionFilter, IResultFilter, I
         }
 
         protected virtual CacheItem GetCacheItem(string key) {
-            try {
-                var cacheItem = _cacheStorageProvider.GetCacheItem(key);
-                return cacheItem;
+            TaskCompletionSource<CacheItem> tcs;
+            if (!_state.Cachers.TryGetValue(key, out tcs)) {
+                tcs = new TaskCompletionSource<CacheItem>();
+
+                // If the key could not be added it means another thread got there first,
+                // and in this case they will both try to get the cache which is fine.
+                // The chances are very low however as even at it needs to happen between two 
+                // statements, even at a high RPS it would be uncommon.
+                _state.Cachers.TryAdd(key, tcs);
+
+                Logger.Debug("Fetching cache item: {0}", _cacheKey);
+
+                try {
+                    var cacheItem = _cacheStorageProvider.GetCacheItem(key);
+                    _state.Cachers.TryRemove(_cacheKey, out tcs);
+                    tcs.SetResult(cacheItem);
+                    return cacheItem;
+                }
+                catch (Exception e) {
+                    Logger.Error(e, "An unexpected error occured while reading a cache entry");
```

Comment:
If you catch the base Exception at least add an `e.IsFatal()` check. BTW aren't we rather conventionally using the variable name `ex` for exceptions, `e` being something common for event object in event handlers?